### PR TITLE
Bugfix/2096 requestbody references

### LINKED ISSUE #2096: OAS3 requestBody $ref format validation is overzealous, gives false positives
### Q&A (please complete the following information)
 - OS: Windows 10
 - Browser: Chrome latest
 - Method of installation: https://editor.swagger.io
 - Swagger-Editor version: 3.8.0
 - Swagger/OpenAPI version: OpenAPI 3.0

### Content & configuration
Example Swagger/OpenAPI definition:
```yaml
openapi: 3.0.2
info:
  title: requestBody $ref test
  version: 1.0.0
paths:
  /foo:
    post:
      requestBody:
        $ref: 'https://httpbin.org/anything/foo.yaml'
      responses:
        '200':
          description: ok
```

### Describe the bug you're encountering

https://github.com/swagger-api/swagger-editor/blob/master/src/plugins/validate-semantic/validators/oas3/refs.js

The `validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions` function here is buggy. It expects that the `requestBody.$ref` path contains the `#/...` part, and that the `#/...` part contains either the `requestBodies` or `requestBody` segment. This is not necessarily the case. For example, a `$ref` may point to an external file with no `#/...` part.

I think the logic should be changed to something like:
```
if ($ref value begins with '#/components'
    and
    $ref value does not begin with '#/components/requestBodies/')
then error
otherwise no error
```
(This is just an example; this doesn't account for the rare/edge cases listed below.)

### Expected behavior
The definition above should not produce errors in the editor.

### Additional context or thoughts
In the context of `requestBody`, the following `$ref` paths are valid:

1\)
```yaml
$ref: '#/components/requestBodies/addPetBody'
```

2\) Any external `$ref` is semantically valid here. There's no way to tell from the `$ref` path alone whether or not such a `$ref` points to a RequestBody Object.
```yaml
$ref: 'addPetBody.yaml'

$ref: './components.yaml#/path/to/some/node'

$ref: 'https://example.com'

$ref: 'https://example.com/file.yaml#/requestBody1'

$ref: 'https://example.com/file.yaml#/components/addPetBody'

$ref: 'https://example.com/file.yaml#/components/requestBodies/addPetBody'

$ref: 'https://example.com/file.yaml#/components/requestBodies/group1/addPetBody'

$ref: 'https://example.com/file.yaml#/requestBodies/addPetBody'
```

3\) Also valid, but rare/edge cases:
```yaml
# refs directly to the `requestBody` node in other paths and callbacks
$ref: '#/paths/~1foo/post/requestBody'

$ref: '#/components/callbacks/callback1/%7B%24request.body%23~1callbackUrl%7D/post/requestBody'
                                        ^^ {$request.body#/callbackUrl} ^^

$ref: '#/paths/~1sfoo/post/callbacks/callback1/%7B%24request.body%23~1callbackUrl%7D/post/requestBody'


# refs where any of the path nodes is an extension
$ref: '#/x-body1'

$ref: '#/x-extension/addPetBody'

$ref: '#/<any-key-that-allows-extensions>/x-extension/body1'

$ref: '#/components/x-body1'
```

NOT valid:
```yaml
$ref: '#/components/schemas/Pet'

$ref: '#/components/schemas/requestBody'
# a schema named "requestBody" is not actually a Request Body component

$ref: '#/paths/~1foo/post/requestBody/schema'
```

### PR DESCRIPTION
**Pull request by ralphdoe**
This resolves some false positives on requestbody refs validations.  
### Description

I added validations for cases specified on ticket #2096.

### Motivation and Context

It fixes #2096 

### How Has This Been Tested?

Added Unit test cases.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [X] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [X] are not breaking changes.

### Documentation
- [X] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [X] All new and existing tests passed.

### CODE REVIEW

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 1]
```diff
@@ -6,26 +6,58 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
         const value = node.node
         const ref = value.$ref
 
-        if(!ref) {
+        if (!ref) {
           return acc
         }
 
         const [, refPath = ""] = ref.split("#")
```
Comment (shockey) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: **change request:** use `refUrl` and `refPath` to feed the logic below this line, instead of using `ref` directly!

this gives us a higher-level view of what the ref actually is, which will make the logic below cleaner IMO. some usage examples:

```js
// a local ref
const [refUrl, refPath = ""] = "#/components/schemas/Pet".split("#")
console.log(refUrl) // ""
console.log(refPath) // "/components/schemas/Pet"

// a remote document ref
const [refUrl, refPath = ""] = "http://google.com/openapi.yaml".split("#")
console.log(refUrl) // "http://google.com/openapi.yaml"
console.log(refPath) // ""

// a remote path ref
const [refUrl, refPath = ""] = "http://google.com/openapi.yaml#/mySchema".split("#")
console.log(refUrl) // "http://google.com/openapi.yaml"
console.log(refPath) // "/mySchema"
```

```suggestion
        const [refUrl, refPath = ""] = ref.split("#")
```

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 2]
```diff
@@ -6,26 +6,58 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+        // External references ending with yaml
+        if (ref.endsWith(".yaml")) {
+          return acc
+        }
+
+        // Starting with #/compontents/schemas is not allowed
+        if (ref.startsWith("#/components/schemas")) {
+          acc.push({
+            level: "error",
+            message: `requestBody $refs cannot point to '#/components/schemas/…', they must point to '#/components/requestBodies/…'`,
```
[the remark this answers is not in the document]
Comment (hkosova) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: @shockey it was my request to add a separate message for this case because it's more common than others. "... position where a requestBody can be legally placed" just doesn't seem a clear enough indication of what's wrong in this particular case. It's better to be specific than vague.

Better wording suggestions are welcome!

> we usually don't say what the user has done wrong in our error messages _and_ how to fix it — it's generally one or the other.

A [good error message](https://hackernoon.com/what-makes-a-good-error-710d02682a68) tells both what's wrong _and_ how to fix it. Some errors are simple enough so that either the "what's wrong" or "how to fix" part is enough. Others could use a more detailed explanation. Fyi I also used the "what's wrong _and_ how to fix" pattern for the error messages [here](https://github.com/swagger-api/swagger-editor/pull/2105) and [here](https://github.com/swagger-api/swagger-editor/pull/2108).

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 3]
```diff
@@ -6,26 +6,58 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+
+        // Edge cases
+        if (ref.endsWith("requestBody") && (ref.startsWith("#/paths") || ref.startsWith("#/components"))){
+          return acc  
+        }
+
+        // External references ending with yaml
+        if (ref.endsWith(".yaml")) {
+          return acc
+        }
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: @ralphdoe all checks for external $refs (`http(s)://`, `./`, `.yaml`) can be replaced with
```js
// Ignore external references
if (!ref.startsWith("#/") {
  return acc
}
```
or
```js
if (ref.startsWith("#/")) {
  // validations for local $refs go here...
  if (...) {...}
}
```
This logic will automatically exclude all non-local $refs such as
```
http(s)://...
./path/to/...
../path/to/...
file.json
subfolder/file.yaml
etc.
```

File: src/plugins/validate-semantic/validators/oas3/refs.js [hunk 4]
```diff
@@ -6,26 +6,58 @@ export const validateOAS3RefsForRequestBodiesReferenceRequestBodyPositions = ()
...
+          return acc  
+        }
+
+        // External references starting with ./
+        if (ref.startsWith("./")){
+          return acc  
+        }
+
+        // Extensions are valid
+        if (ref.startsWith("#/") && (parentRefKey.startsWith("/x-") || targetRefKey.startsWith("/x-"))){
```
Comment (hkosova) [inline src/plugins/validate-semantic/validators/oas3/refs.js]: @ralphdoe 
1) `startsWith("/x-")` -->`startsWith("x-")`  (i.e. no slash)
2) All items of `pathArr` need to be checked for `x-`, not just `parentRefKey` and `targetRefKey`. The logic should be
```
if local $ref
and
any item in pathArr starts with "x-"
```

File: test/unit/plugins/validate-semantic/oas3/refs.js [hunk 5]
```diff
@@ -53,10 +53,7 @@ describe("validation plugin - semantic - oas3 refs", () => {
       return validateHelper(spec)
         .then(system => {
           const allErrors = system.errSelectors.allErrors().toJS()
-          const firstError = allErrors[0]
-          expect(allErrors.length).toEqual(1)
-          expect(firstError.message).toEqual(`requestBody $refs must point to a position where a requestBody can be legally placed`)
-          expect(firstError.path).toEqual(["paths", "/", "post", "requestBody", "$ref"])
+          expect(allErrors.length).toEqual(0)
```
[the remark this answers is not in the document]
Comment (ralphdoe) [reply test/unit/plugins/validate-semantic/oas3/refs.js]: Done!

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 3 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Done.

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 4 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: Done.

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 1 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: I specified some refPath, but as we are ignoring all external references. I don't know where to use refUrl.

File: src/plugins/validate-semantic/validators/oas3/refs.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/oas3/refs.js]: I left the message as we discussed with Helen, if there's a better message or makes sense let it as it was, i can do it.
